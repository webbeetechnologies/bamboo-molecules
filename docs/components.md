# Bamboo Molecules

Bamboo Molecules components are built on top of BambooAtoms; thus, at it's core; Bamboo allows for all the components to be composed by default. That means; you can replace an underlying view by a Styled View of your choice; and that is only the beginning.


Bamboo atoms have minimal theming; and Bamboo Molecules are opinionated styles; however, Molecules allow you to customize or entirely change the theme of a specific component. It is possible to use custom components/styles at a module level.


Molecules is heavily inspired by Material and Cupertino components and thus tries to mimic them to provide a native experience as closely as possible on the react native platform.



# References:
- [Comparing iOS and Android UI](https://www.learnui.design/blog/ios-vs-android-app-ui-design-complete-guide.html)
- [Material 3 Figma Design](https://www.figma.com/file/hGlWM0wnP1UN5VtPGiaOJ2/Material-3-Design-Kit-(Community)?node-id=50716%3A11360)
- [Material 3 Design Guidelines](https://m3.material.io/components)

## Molecules HOC:
- [WithAddons](./components/HOC/WithAddons.md)
- [WithAddonAfter](./components/HOC/WithAddons.md#withaddonafter)
- [WithAddonBefore](./components/HOC/WithAddons.md#withaddonbefore)
- [WithElements](./components/HOC/WithElements.md)
- [WithActionIconButton](./components/HOC/WithActionIconButton.md)
- [WithPopover](./components/HOC/WithPopover.md)


## Molecules created using Atoms:
- High Priority
    -   [Icons](./components/Icons.md)
    -   [Input](./components/Input.md)
    -   [Tooltips](./components/Tooltips.md)
    -   Actionsheet
    -   [Flat List](./components/FlatList.md)
    -   Modals/ Dialogs
    -   [Button](./components/Button.md)
    -   [IconButton](./components/IconButton.md)
    -   [Select](./components/Select.md)
    -   [Radio](./components/Radio.md)
    -   [Checkbox](./components/Checkbox.md)
    -   [Date Picker](./components/DatePicker.md)
    -   [Time Picker](./components/TimePicker.md)
    -   Menu and Menu Items
    -   Alert
    -   Cards
    -   Accordion
    -   [Chips](./components/Chips.md)
    -   Drawer


- Medium Priority
    -   [Section List](./components/SectionList.md)
    -   [Range Picker](./components/RangePickers.md)
    -   [Date Time Picker](./components/DateTimePicker.md)
    -   Stack
    -   Divider
    -   Toast
    -   Fab/ Fab Group (Android/Web Only)
    -   Tabs
        -   Pills
        -   Bottom Tabs


- Low Priority
    -   [Color Picker](./components/ColorPicker.md)
    -   [FilePicker](./components/FilePicker.md)
    -   Badges
    -   Search
    -   Forms
    -   Avatar
    -   Graphs




... and more as we identify ..
