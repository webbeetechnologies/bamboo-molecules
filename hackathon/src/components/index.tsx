export { default as AttributeValueInput } from './AttributeValueInput/AttributeValueInput';
export { default as CustomDrawer } from './CustomDrawer/CustomDrawer';
export { default as MachineCard } from './MachineCard/MachineCard';
export { default as MachineTypeCard } from './MachineTypeCard/MachineTypeCard';
