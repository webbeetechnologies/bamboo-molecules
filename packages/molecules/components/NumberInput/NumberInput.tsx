import { forwardRef, memo, useMemo } from 'react';
import type { TextInputProps } from '../TextInput';
import { createNumberMask } from '../../utils';
import { MaskedInput } from '../MaskedInput';

export type Props = Omit<TextInputProps, 'value' | 'defaultValue' | 'onChangeText'> & {
    /**
     * required for all the maskedinputs
     * */
    value: string;
    /**
     * required for all the maskedinput
     * */
    onChangeText: (value: string) => void;
    /**
     * Number-only keyboardType
     */
    keyboardType?: 'number-pad' | 'decimal-pad' | 'numeric' | 'phone-pad';
    /**
     * Character for thousands delimiter. Defaults to `"."`
     * */
    delimiter?: string;
    /**
     * Decimal precision. Defaults to `2`
     * */
    precision?: number;
    /**
     * Decimal separator character. Defaults to `","`
     * */
    separator?: string;
    /**
     * If this is false, precision will be ignored
     * */
    allowDecimals?: boolean;
    /**
     * For default prefix, if we want to allow negative or not
     * If the custom prefix is passed, this will be ignored.
     * */
    allowNegative?: boolean;
    /**
     * Mask to be prefixed on the mask result
     * */
    prefix?: string | string[];
    /**
     * Mask to be suffixed on the mask result
     * */
    suffix?: string | string[];
};

const NumberInput = (
    {
        keyboardType = 'numeric',
        delimiter = ',',
        precision = 2,
        separator = '.',
        allowDecimals = true,
        allowNegative = true,
        prefix = [],
        suffix = '',
        ...rest
    }: Props,
    ref: any,
) => {
    const mask = useMemo(
        () =>
            createNumberMask({
                delimiter,
                precision: !allowDecimals ? 0 : precision,
                separator,
                prefix,
                allowNegative,
                suffix,
            }),
        [allowDecimals, allowNegative, delimiter, precision, prefix, separator, suffix],
    );

    return <MaskedInput mask={mask} {...rest} keyboardType={keyboardType} ref={ref} />;
};

export default memo(forwardRef(NumberInput));
