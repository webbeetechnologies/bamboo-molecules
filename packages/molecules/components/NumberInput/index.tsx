import { getRegisteredComponentWithFallback, registerMoleculesComponents } from '../../core';
import NumberInputDefault from './NumberInput';

registerMoleculesComponents({
    NumberInput: NumberInputDefault,
});

export { Props as NumberInputProps } from './NumberInput';

export const NumberInput = getRegisteredComponentWithFallback('NumberInput', NumberInputDefault);
