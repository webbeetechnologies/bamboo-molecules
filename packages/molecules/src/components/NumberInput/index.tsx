import { getRegisteredComponentWithFallback, registerMoleculesComponents } from '../../core';
import NumberInputDefault from './NumberInput';

registerMoleculesComponents({
    NumberInput: NumberInputDefault,
});

export const NumberInput = getRegisteredComponentWithFallback('NumberInput', NumberInputDefault);
