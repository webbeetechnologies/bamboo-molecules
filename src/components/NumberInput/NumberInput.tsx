import { forwardRef, memo, useMemo } from 'react';
import type { MaskArray } from 'react-native-mask-input';

import { useMolecules } from '../../hooks';
import type { TextInputProps } from '../TextInput';
import { createNumberMask } from '../../utils';

export type Props = Omit<TextInputProps, 'value' | 'defaultValue' | 'onChangeText'> & {
    /**
     * required for all the maskedinputs
     * */
    value: string;
    /**
     * required for all the maskedinput
     * */
    onChangeText: (value: string) => void;
    /**
     * Number-only keyboardType
     */
    keyboardType?: 'number-pad' | 'decimal-pad' | 'numeric' | 'phone-pad';
    /**
     * Character for thousands delimiter. Defaults to `"."`
     * */
    delimiter?: string;
    /**
     * Decimal precision. Defaults to `2`
     * */
    precision?: number;
    /**
     * Decimal separator character. Defaults to `","`
     * */
    separator?: string;
    /**
     * If this is false, precision will be ignored
     * */
    allowDecimals?: boolean;
    /**
     * For default prefix, if we want to allow negative or not
     * If the custom prefix is passed, this will be ignored.
     * */
    allowNegative?: boolean;
    /**
     * Mask to be prefixed on the mask result
     * */
    prefix?: MaskArray;
    /**
     * Mask to be suffixed on the mask result
     * */
    suffix?: string | string[];
};

const NumberInput = (
    {
        keyboardType = 'numeric',
        delimiter = '.',
        precision = 2,
        separator = '',
        allowDecimals = true,
        allowNegative = true,
        prefix,
        suffix,
        ...rest
    }: Props,
    ref: any,
) => {
    const { MaskedInput } = useMolecules();
    const mask = useMemo(
        () =>
            createNumberMask({
                delimiter,
                precision: !allowDecimals ? 0 : precision,
                separator,
                prefix: prefix ? prefix : allowNegative ? [/^[\d+-]+$/] : [/^[\d+]+$/],
                suffix,
            }),
        [allowDecimals, allowNegative, delimiter, precision, prefix, suffix, separator],
    );

    return <MaskedInput mask={mask} {...rest} keyboardType={keyboardType} ref={ref} />;
};

export default memo(forwardRef(NumberInput));
