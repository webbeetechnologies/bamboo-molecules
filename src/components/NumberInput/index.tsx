export { default as NumberInput, Props as NumberInputProps, NumberInputMasks } from './NumberInput';
