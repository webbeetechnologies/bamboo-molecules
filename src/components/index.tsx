export {
    TextProps,
    HeadingProps,
    H1Props,
    H2Props,
    H3Props,
    H4Props,
    H6Props,
    UnderlineProps,
    StrongProps,
    LabelProps,
    StrikethroughProps,
    ItalicProps,
    ViewProps,
} from '@webbee/bamboo-atoms';

export { Icon, registerCustomIconType, IconProps, iconStyles } from './Icon';
export { TouchableRipple, TouchableRippleProps, touchableRippleStyles } from './TouchableRipple';
export {
    ActivityIndicator,
    ActivityIndicatorProps,
    activityIndicatorStyles,
} from './ActivityIndicator';
export { Surface, SurfaceProps, surfaceStyles } from './Surface';
export { Button, ButtonProps, buttonStyles } from './Button';
export {
    HorizontalDivider,
    horizontalDividerStyles,
    HorizontalDividerProps,
} from './HorizontalDivider';
export { VerticalDivider, verticalDividerStyles, VerticalDividerProps } from './VerticalDivider';
export { Switch, switchStyles, SwitchProps } from './Switch';
export { FlatList, FlatListProps, IFlatList, flatListStyles } from './FlatList';

export {
    SectionList,
    SectionListProps,
    ISectionList,
    SectionListRenderItem,
    SectionListRenderItemInfo,
    DefaultSectionT,
    sectionListStyles,
} from './SectionList';

export {
    default as ListItem,
    ListItemProps,
    listItemStyles,
    listItemTitleStyles,
    listItemDescriptionStyles,
} from './ListItem';

export { FlatGrid, IFlatGrid, flatGridStyles, FlatGridProps } from './FlatGrid';

export { SectionGrid, ISectionGrid, sectionGridStyles, SectionGridProps } from './SectionGrid';

export { Popover, PopoverProps, popoverStyles } from './Popover';
export { Overlay, OverlayProps } from './Overlay';
export { Backdrop, BackdropProps, backdropStyles } from './Backdrop';
export * from './Animations';

export { IconButton, IconButtonProps, iconButtonStyles } from './IconButton';
export { default as Checkbox, CheckboxProps, checkboxStyles } from './Checkbox';
export { TextInput, TextInputProps, textInputStyles } from './TextInput';
export { NumberInput, NumberInputProps } from './NumberInput';
export { FilePicker, FilePickerProps } from './FilePicker';
export { HelperText, HelperTextProps, helperTextStyles } from './HelperText';

export { Modal, ModalProps, modalStyles } from './Modal';

export {
    DatePickerInline,
    DatePickerInlineProps,
    datePickerStyles,
    datePickerYearPickerStyles,
    datePickerMonthStyles,
    datePickerDayStyles,
    datePickerHeaderStyles,
    dateDayNameStyles,
    datePickerDayRangeStyles,
    datePickerDayEmptyStyles,
    datePickerWeekStyles,
    CalendarDate,
    CalendarDates,
} from './DatePickerInline';

export {
    DatePickerModal,
    DatePickerModalProps,
    datePickerModalStyles,
    datePickerModalContentHeaderStyles,
    datePickerModalHeaderStyles,
    datePickerModalHeaderBackgroundStyles,
    datePickerModalEditStyles,
} from './DatePickerModal';

export { DatePickerInput, DatePickerInputProps, datePickerInputStyles } from './DatePickerInput';

export {
    TimePicker,
    TimePickerProps,
    timePickerStyles,
    timePickerInputStyles,
    timePickerInputsStyles,
    timePickerClockStyles,
    timePickerClockHoursStyles,
    timePickerClockMinutesStyles,
    timePickerAmPmSwitcherStyles,
} from './TimePicker';

export { TimePickerModal, TimePickerModalProps, timePickerModalStyles } from './TimePickerModal';

export { TextInputWithMask, TextInputWithMaskProps } from './TextInputWithMask';

export { OptionList, OptionListProps, IOptionList, optionListStyles } from './OptionList';

export { DropdownList, DropdownListProps, IDropdownList, dropdownListStyles } from './DropdownList';

export { ElementGroup, elementGroupStyles, ElementGroupProps } from './ElementGroup';

export { InputAddon, inputAddonStyles, InputAddonProps } from './InputAddon';

export * from './ActionSheet';

export * from './Dialog';

export * from './RadioButton';

export * from './NumberRangeInput';

export {
    ProgressCircle,
    ProgressCircleProps,
    progressCircleStyles,
} from './ProgressIndicator/Circle';
export { ProgressBar, ProgressBarProps, progressBarStyles } from './ProgressIndicator/Bar';

export * from './DropdownListPopover';

export * from './Card';

export * from './Select';

export * from './Chip';
