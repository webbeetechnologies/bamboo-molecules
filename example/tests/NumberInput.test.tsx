import { renderWithWrapper } from '../testHelpers';
import { NumberInput } from '../../src/components/NumberInput';

const mockedOnChangeText = jest.fn();

it('should render correctly with the default mask', () => {
    const tree = renderWithWrapper(
        <NumberInput
            value="with space and special* characters;"
            onChangeText={mockedOnChangeText}
        />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('should correctly mask default value', () => {
    const { getByDisplayValue } = renderWithWrapper(
        <NumberInput defaultValue="-123.22abc" onChangeText={mockedOnChangeText} />,
    );
    expect(getByDisplayValue('-123.22')).toBeTruthy();
});
