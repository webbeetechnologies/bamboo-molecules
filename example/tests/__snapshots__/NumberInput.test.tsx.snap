// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with the default mask 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "animationScale": 1,
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
        "flexDirection": "row",
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      style={
        Object {
          "flexGrow": 1,
          "minHeight": 56,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <TextInput
        editable={true}
        keyboardType="numeric"
        maxFontSizeMultiplier={15}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="rgba(73, 69, 79, 1)"
        size="md"
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "flexGrow": 1,
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
              "paddingBottom": 8,
              "paddingTop": 24,
            },
            Object {
              "height": 56,
            },
            false,
            Object {
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
</View>
`;
